lib/Sema/CSDiagnostics.cpp

```diff
@@ -981,100 +981,128 @@ bool AssignmentFailure::diagnoseAsError() {
-    else if (VD->isCaptureList())
-      message += " is an immutable capture";
-    else if (VD->isImplicit())
-      message += " is immutable";
-    else if (VD->isLet())
-      message += " is a 'let' constant";
-    else if (!VD->isSettable(DC))
-      message += " is a get-only property";
-    else if (!VD->isSetterAccessibleFrom(DC))
-      message += " setter is inaccessible";
-    else {
-      message += " is immutable";
-    }
-
-    emitDiagnostic(Loc, DeclDiagnostic, message)
-        .highlight(immInfo.first->getSourceRange());
-
-    // If there is a masked instance variable of the same type, emit a
-    // note to fixit prepend a 'self.'.
-    if (auto typeContext = DC->getInnermostTypeContext()) {
-      UnqualifiedLookup lookup(VD->getFullName(), typeContext,
-                               getASTContext().getLazyResolver());
-      for (auto &result : lookup.Results) {
-        const VarDecl *typeVar = dyn_cast<VarDecl>(result.getValueDecl());
-        if (typeVar && typeVar != VD && typeVar->isSettable(DC) &&
-            typeVar->isSetterAccessibleFrom(DC) &&
-            typeVar->getType()->isEqual(VD->getType())) {
-          // But not in its own accessor.
-          auto AD =
-              dyn_cast_or_null<AccessorDecl>(DC->getInnermostMethodContext());
-          if (!AD || AD->getStorage() != typeVar) {
-            emitDiagnostic(Loc, diag::masked_instance_variable,
-                           typeContext->getSelfTypeInContext())
-                .fixItInsert(Loc, "self.");
+  Optional<OverloadChoice> choice = immInfo.second;
+
+  // Attempt diagnostics based on the overload choice.
+  if (choice.hasValue()) {
+
+    if (!choice->isDecl()) {
+      if (choice->getKind() == OverloadChoiceKind::KeyPathApplication &&
+          !isa<ApplyExpr>(immInfo.first)) {
+        std::string message = "the key path";
+        if (auto *SE = dyn_cast<SubscriptExpr>(immInfo.first)) {
+          if (auto *tupleExpr = dyn_cast<TupleExpr>(SE->getIndex())) {
+            if (auto *DRE = dyn_cast<DeclRefExpr>(tupleExpr->getElement(0))) {
+              auto identifier = DRE->getDecl()->getBaseName().getIdentifier();
+              message = "'" + identifier.str().str() + "'";
+            }
           }
         }
+        emitDiagnostic(Loc, DeclDiagnostic, message + " is a read-only key path")
+            .highlight(immInfo.first->getSourceRange());
+        return true;
+      }
+      return false;
+    } 
+
+    // Otherwise, we cannot resolve this because the available setter candidates
+    // are all mutating and the base must be mutating.  If we dug out a
+    // problematic decl, we can produce a nice tailored diagnostic.
+    if (auto *VD = dyn_cast_or_null<VarDecl>(choice->getDecl())) {
+      std::string message = "'";
+      message += VD->getName().str().str();
+      message += "'";
+
+      auto type = getType(immInfo.first);
+
+      if (isKnownKeyPathType(type))
+        message += " is a read-only key path";
```

Comment:
You could probably do this by adding `select` to the diagnostic message, although it might be overkill if you have so many possibilities here.

Reply:
That select unfortunately is going to be pretty big, I'd suggest to drop trailing "key path" in this case though.